### ISSUE TITLE
java.lang.AssertionError: assertion failed: ClassBType.info not yet assigned

### ISSUE THREAD

**Issue by Diagoras** [wix/accord#103]
Receiving this exception in my project when compiling with 2.12.1

Full stack trace:

`java.lang.AssertionError: assertion failed: ClassBType.info not yet assigned: Lcom/wix/accord/transform/ValidationTransform$TransformedValidator;
	at scala.tools.nsc.backend.jvm.BTypes$ClassBType.info(BTypes.scala:861)
	at scala.tools.nsc.backend.jvm.BTypes$ClassBType.isInterface(BTypes.scala:904)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genCallMethod(BCodeBodyBuilder.scala:1073)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:621)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:298)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:368)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:68)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.$anonfun$genBlock$1(BCodeBodyBuilder.scala:812)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlock(BCodeBodyBuilder.scala:812)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:364)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:603)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:635)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:509)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.$anonfun$gen$7(BCodeSkelBuilder.scala:511)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:511)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:113)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase$Worker1.visit(GenBCode.scala:190)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase$Worker1.$anonfun$run$1(GenBCode.scala:139)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase$Worker1.run(GenBCode.scala:139)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase.buildAndSendToDisk(GenBCode.scala:381)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase.run(GenBCode.scala:350)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1418)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1403)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1398)
	at scala.tools.nsc.Global$Run.compile(Global.scala:1492)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:123)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:99)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:102)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:48)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:41)
	at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply$mcV$sp(AggressiveCompile.scala:99)
	at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply(AggressiveCompile.scala:99)
	at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply(AggressiveCompile.scala:99)
	at sbt.compiler.AggressiveCompile.sbt$compiler$AggressiveCompile$$timed(AggressiveCompile.scala:166)
	at sbt.compiler.AggressiveCompile$$anonfun$3.compileScala$1(AggressiveCompile.scala:98)
	at sbt.compiler.AggressiveCompile$$anonfun$3.apply(AggressiveCompile.scala:143)
	at sbt.compiler.AggressiveCompile$$anonfun$3.apply(AggressiveCompile.scala:87)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:39)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:37)
	at sbt.inc.IncrementalCommon.cycle(Incremental.scala:99)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:38)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:37)
	at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:65)
	at sbt.inc.Incremental$.compile(Incremental.scala:37)
	at sbt.inc.IncrementalCompile$.apply(Compile.scala:27)
	at sbt.compiler.AggressiveCompile.compile2(AggressiveCompile.scala:157)
	at sbt.compiler.AggressiveCompile.compile1(AggressiveCompile.scala:71)
	at com.typesafe.zinc.Compiler.compile(Compiler.scala:184)
	at com.typesafe.zinc.Compiler.compile(Compiler.scala:164)
	at sbt_inc.SbtIncrementalCompiler.compile(SbtIncrementalCompiler.java:92)
	at scala_maven.ScalaCompilerSupport.incrementalCompile(ScalaCompilerSupport.java:303)
	at scala_maven.ScalaCompilerSupport.compile(ScalaCompilerSupport.java:119)
	at scala_maven.ScalaCompilerSupport.doExecute(ScalaCompilerSupport.java:99)
	at scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:482)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Error while emitting Validations.scala
assertion failed: ClassBType.info not yet assigned: Lcom/wix/accord/transform/ValidationTransform$TransformedValidator;`

**Comment by holograph** [comment]
Thanks for the report, I'll look into this as soon as possible.

**Comment by Diagoras** [comment]
Thanks for the quick followup. I'm using version 0.6.1 of Accord, let me know if you need any other information (content of failing classes, etc.)

**Comment by helllamer** [comment]
Same problem for `accord 0.7-SNAPSHOT` + `scala-2.12.1`.

Steps to reproduce:
$ `git clone https://github.com/scala-js/scalajs-tutorial`
$ `cd scalajs-tutorial`
```
cat >> build.sbt << EOF

resolvers += "Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"

libraryDependencies += "com.wix" %%% "accord-core" % "0.7-SNAPSHOT"

EOF
```

```
cat >> src/main/scala/tutorial/webapp/A.scala << EOF
case class A(v: Int)

object A {
  import com.wix.accord.dsl._
  implicit val VALIDATOR = validator[A] { a =>
    a.v should be >= 1
  }
}
EOF
```
$ `sbt fullOptJS`
It returns this:
```
[info] Updating {file:/mnt/src/user/projects/repo/scalajs-tutorial/}scalajs-tutorial...
[info] Resolving com.wix#accord-core_sjs0.6_2.12;0.7-SNAPSHOT ...
[warn] Unable to reparse com.wix#accord-core_sjs0.6_2.12;0.7-SNAPSHOT from Sonatype OSS Snapshots, using Wed Jan 04 18:29:11 MSK 2017
[info] Resolving com.wix#accord-api_sjs0.6_2.12;0.7-SNAPSHOT ...
[warn] Unable to reparse com.wix#accord-api_sjs0.6_2.12;0.7-SNAPSHOT from Sonatype OSS Snapshots, using Wed Jan 04 18:28:54 MSK 2017
[info] Resolving org.eclipse.jetty#jetty-continuation;8.1.16.v20140903 ...
[info] Done updating.

[info] Compiling 1 Scala source to /mnt/src/user/projects/repo/scalajs-tutorial/target/scala-2.12/classes...
[error] missing or invalid dependency detected while loading class file 'ValidationTransform.class'.
[error] Could not access term universe in class scala.reflect.macros.blackbox.Context,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'ValidationTransform.class' was compiled against an incompatible version of scala.reflect.macros.blackbox.Context.
[error] missing or invalid dependency detected while loading class file 'MacroHelper.class'.
[error] Could not access type Context in package scala.reflect.macros.blackbox,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'MacroHelper.class' was compiled against an incompatible version of scala.reflect.macros.blackbox.
java.lang.IllegalArgumentException: requirement failed: package parser
	at scala.reflect.internal.Types$ModuleTypeRef.<init>(Types.scala:1879)
	at scala.reflect.internal.Types$PackageTypeRef.<init>(Types.scala:1897)
	at scala.reflect.internal.Types$TypeRef$.apply(Types.scala:2401)
	at scala.reflect.internal.Types.typeRef(Types.scala:3553)
	at scala.reflect.internal.Types.typeRef$(Types.scala:3536)
	at scala.reflect.internal.SymbolTable.typeRef(SymbolTable.scala:16)
	at scala.reflect.internal.Symbols$TypeSymbol.newTypeRef(Symbols.scala:3026)
	................
[error] Error while emitting TutorialApp.scala
[error] requirement failed: package parser
java.lang.AssertionError: assertion failed: ClassBType.info not yet assigned: Ltutorial/webapp/A$;
	at scala.tools.nsc.backend.jvm.BTypes$ClassBType.info(BTypes.scala:861)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.initJClass(BCodeSkelBuilder.scala:129)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:98)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase$Worker1.visit(GenBCode.scala:190)
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase$Worker1.$anonfun$run$1(GenBCode.scala:139)
        ............
```

**Comment by helllamer** [comment]
I've tried to "full rebuild" (as recomended by `scalac`):
rebuilt `accord` from git against scala-2.12.1, and `publish-local`'ed it.
This does not help, home-made jars produces same error.

**Comment by holograph** [comment]
I'll have some time to give this a serious look over the weekend. Will keep
you posted.

**Comment by holograph** [comment]
Apologies for the delay, I've been quite ill these last few weeks and haven't had much time to work on Accord. I've now reproduced the problem with Maven deps and am looking into this.

**Comment by holograph** [comment]
Verified on all counts, but I hit a brick wall in trying to figure it out. I opened a Scala compiler bug (see [SI-10171](https://issues.scala-lang.org/browse/SI-10171)) and will keep this issue up to date with developments.

I apologize for what must be a very frustrating blocker to upgrading Scala for some of you; I only hope the Scala team can assist with an expedited fix or workaround.

**Comment by holograph** [comment]
Will review with the next Scala compiler release; I'm keeping this for the 0.7 milestone to remind myself to update the documentation.

**Comment by holograph** [comment]
Should be resolved as of 2.12.2.
